[view-transitions] Add basic DOM infrastructure

nt1m opened the pull request:
#### b10105f3d6b7e01e12f1c141a0aebc023f2324c9
<pre>
[view-transitions] Add basic DOM infrastructure
<a href="https://bugs.webkit.org/show_bug.cgi?id=265155">https://bugs.webkit.org/show_bug.cgi?id=265155</a>
<a href="https://rdar.apple.com/118659606">rdar://118659606</a>

Reviewed by Simon Fraser.

Add basic DOM interfaces and lifecycle hooks.

In order to start passing reftests, we need to implement the barebone DOM infrastructure and lifecycle hooks.
We mostly focus on the web exposed methods and managing the promise states, as it is the minimum required to run reftests without having them timeout/throw.
FIXMEs are added to implement the actual functionality.

From <a href="https://drafts.csswg.org/css-view-transitions/#lifecycle">https://drafts.csswg.org/css-view-transitions/#lifecycle</a> :

A successful view transition goes through the following phases:

1. Developer calls document.startViewTransition(updateCallback), which returns a ViewTransition, viewTransition.
2. Current state captured as the “old” state.
3. Rendering paused.
4. Developer’s updateCallback function, if provided, is called, which updates the document state.
5. viewTransition.updateCallbackDone fulfills.
6. Current state captured as the “new” state.
7. Transition pseudo-elements created. See § 3.2 View Transition Pseudo-elements for an overview of this structure.
8. Rendering unpaused, revealing the transition pseudo-elements.
9. viewTransition.ready fulfills.
10. Pseudo-elements animate until finished.
11. Transition pseudo-elements removed.
12. viewTransition.finished fulfills.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-capture-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-start-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/group-animation-for-root-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/input-targets-root-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/mix-blend-mode-only-on-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-crash-set-exception-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-css-animation-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-raf-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-on-root-element-with-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/paused-animation-at-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-computed-style-stays-in-sync-with-new-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/ready_resolves_after_dom_before_raf-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/style-inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/synchronous-callback-skipped-before-run-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-after-animation-started-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-from-invalid-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/unset-and-initial-view-transition-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-name-on-removed-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/web-animation-pseudo-incorrect-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document+ViewTransition.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::activeViewTransition const):
(WebCore::Document::setActiveViewTransition):
(WebCore::Document::hasViewTransitionPseudoElementTree const):
(WebCore::Document::setHasViewTransitionPseudoElementTree):
(WebCore::Document::startViewTransition):
(WebCore::Document::performPendingViewTransitions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp: Added.
(WebCore::createPromiseAndWrapper):
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::ready):
(WebCore::ViewTransition::updateCallbackDone):
(WebCore::ViewTransition::finished):
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::skipTransition):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::clearViewTransition):
* Source/WebCore/dom/ViewTransition.h: Added.
(WebCore::ViewTransition::phase):
(WebCore::ViewTransition::namedElements const):
(WebCore::ViewTransition::protectedDocument const):
* Source/WebCore/dom/ViewTransition.idl: Added.
* Source/WebCore/dom/ViewTransitionUpdateCallback.h: Added.
* Source/WebCore/dom/ViewTransitionUpdateCallback.idl: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):

Canonical link: <a href="https://commits.webkit.org/271426@main">https://commits.webkit.org/271426@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d946e911cffafd5d74d64a9162cc9db2a8f516e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28360 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7004 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30887 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25818 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9097 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4373 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24399 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/5046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25398 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31573 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25970 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25835 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3287 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6706 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6794 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5562 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5626 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -1844,6 +1846,12 @@ class Document
 
     void detachFromFrame();
 
+    Ref<ViewTransition> startViewTransition(RefPtr<ViewTransitionUpdateCallback> = nullptr);
```
These might be better moved up in the public section, if there's some logical grouping with other functions. Don't always just add stuff at the end.

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -2482,6 +2490,9 @@ class Document
 
     RefPtr<ResizeObserver> m_resizeObserverForContainIntrinsicSize;
+
+    RefPtr<ViewTransition> m_activeViewTransition;
```
Ditto.

smfr commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -0,0 +1,103 @@
+    ViewTransitionPhase phase() { return m_phase; }
+
+private:
+    Ref<Document> m_document;
```
You've got a retain cycle here between Document and ViewTransition. You'll need to ensure it's always explicitly broken somehow.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9787,6 +9789,42 @@ void Document::updateContentRelevancyForScrollIfNeeded(const Element& scrollAnch
+{
+    Ref viewTransition = ViewTransition::create(*this, WTFMove(updateCallback));
+
+    if (m_activeViewTransition)
```
```
if (RefPtr activeViewTransition = m_activeViewTransition)
    activeViewTransition->skipViewTransition(Exception { ExceptionCode::AbortError, "Old view transition aborted by new view transition."_s });
```

This is our modern smart pointer practice. You need to make sure you have a smart pointer *on the stack* when you call a non-trivial member function on it or pass it as argument to a function.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9787,6 +9789,42 @@ void Document::updateContentRelevancyForScrollIfNeeded(const Element& scrollAnch
+{
+    if (!m_activeViewTransition)
+        return;
+    Ref protectedActiveViewTransition { *m_activeViewTransition };
```
We don't use this pattern anymore. This should be:
```
Ref activeViewTransition = *m_activeViewTransition;
```

Then use `activeViewTransition` below instead of `m_activeViewTransition`

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9787,6 +9789,42 @@ void Document::updateContentRelevancyForScrollIfNeeded(const Element& scrollAnch
+    m_activeViewTransition = WTFMove(viewTransition);
+}
+
+Ref<ViewTransition> Document::startViewTransition(RefPtr<ViewTransitionUpdateCallback> updateCallback)
```
`RefPtr<ViewTransitionUpdateCallback>&&`

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -1609,6 +1611,12 @@ class Document
 
+    Ref<ViewTransition> startViewTransition(RefPtr<ViewTransitionUpdateCallback> = nullptr);
+    RefPtr<ViewTransition> activeViewTransition() const;
+    void setActiveViewTransition(RefPtr<ViewTransition>);
```
`RefPtr<ViewTransition>&&`

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -1609,6 +1611,12 @@ class Document
     void resetObservationSizeForContainIntrinsicSize(Element&);
 
+    Ref<ViewTransition> startViewTransition(RefPtr<ViewTransitionUpdateCallback> = nullptr);
+    RefPtr<ViewTransition> activeViewTransition() const;
```
This can return a `ViewTransition*`, this is not a `protectedActiveViewTransition()` function.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    auto deferredPromise = DeferredPromise::create(globalObject);
```
RefPtr

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    auto deferredPromise = DeferredPromise::create(globalObject);
+    auto domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
```
Ref

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+    if (!m_document)
+        return;
+
+    Ref protectedThis { *this };
```
No, we don't do protectors anymore. It is up to the caller of this function to protect `this`.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+    ASSERT(m_phase != ViewTransitionPhase::Done);
+
+    if (m_phase == ViewTransitionPhase::PendingCapture) {
+        protectedDocument()->checkedEventLoop()->queueTask(TaskSource::DOMManipulation, [this]() {
```
What guarantees that `this` is still alive when the lambda runs? Odds are you need to capture a protectedThis or weakThis.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+        }, RejectAsHandled::Yes);
+    }
+
+    m_updateCallbackDone.first->whenSettled([this, protectedThis = Ref { *this }]() {
```
nit: `()` is superfluous.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+
+    ASSERT(m_phase == ViewTransitionPhase::Done || m_phase == ViewTransitionPhase::PendingCapture);
+
+    Ref protectedThis { *this };
```
No protectors in new code please.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+
+    // FIXME: Capture the old state.
+    // FIXME: Set document’s rendering suppression for view transitions to true.
+    protectedDocument()->checkedEventLoop()->queueTask(TaskSource::DOMManipulation, [this]() {
```
missing weakThis.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+        m_phase = ViewTransitionPhase::UpdateCallbackCalled;
+
+    callbackPromise->whenSettled([this, weakThis = WeakPtr { *this }, callbackPromise] () mutable {
+        if (!weakThis)
```
You need to assigned to a RefPtr and then null check. Otherwise, nothing protects `this` during the execution of this lambda:
```
RefPtr protectedThis = weakThis.get();
if (!protectedThis)
    return;
``` 

I know I said "no protectors" but lambda where you capture a weakPtr are an exception, since the "caller" on the lambda is not protecting this.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+
+        callUpdateCallback();
+        m_updateCallbackDone.first->whenSettled([this, weakThis = WeakPtr { *this }]() {
+            if (!weakThis)
```
Missing protectedThis.

cdumez commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -0,0 +1,268 @@
+
+    // FIXME: Implement steps 3-4.
+
+    m_document->setActiveViewTransition(nullptr);
```
`protectedDocument()->setActiveViewTransition(nullptr);`

cdumez commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -0,0 +1,108 @@
+
+class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<ViewTransition> {
+public:
+    ViewTransition(Document&, RefPtr<ViewTransitionUpdateCallback>&&);
```
The constructor needs to be private since this is a RefCounted class with a create() factory function.

cdumez commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -0,0 +1,108 @@
+    ViewTransition(Document&, RefPtr<ViewTransitionUpdateCallback>&&);
+    ~ViewTransition();
+
+    static Ref<ViewTransition> create(Document&, RefPtr<ViewTransitionUpdateCallback>&&);
```
Would be nice to put this one a little higher in the class, like we usually do.

cdumez commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -0,0 +1,108 @@
+    DOMPromise& updateCallbackDone();
+    DOMPromise& finished();
+
+    ViewTransitionPhase phase() { return m_phase; }
```
`phase() const`

cdumez commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -0,0 +1,108 @@
+
+    ViewTransitionPhase phase() { return m_phase; }
+
+    Ref<Document> protectedDocument() const { return *m_document; }
```
I think we've determined that the Document *can* be null, and that's why you used a WeakPtr as data member. As a result, I don't think `*m_document` is safe here.

nt1m committed:
[view-transitions] Add basic DOM infrastructure
https://bugs.webkit.org/show_bug.cgi?id=265155
rdar://118659606

Reviewed by Simon Fraser.

Add basic DOM interfaces and lifecycle hooks.

In order to start passing reftests, we need to implement the barebone DOM infrastructure and lifecycle hooks.
We mostly focus on the web exposed methods and managing the promise states, as it is the minimum required to run reftests without having them timeout/throw.
FIXMEs are added to implement the actual functionality.

From https://drafts.csswg.org/css-view-transitions/#lifecycle :

A successful view transition goes through the following phases:

1. Developer calls document.startViewTransition(updateCallback), which returns a ViewTransition, viewTransition.
2. Current state captured as the “old” state.
3. Rendering paused.
4. Developer’s updateCallback function, if provided, is called, which updates the document state.
5. viewTransition.updateCallbackDone fulfills.
6. Current state captured as the “new” state.
7. Transition pseudo-elements created. See § 3.2 View Transition Pseudo-elements for an overview of this structure.
8. Rendering unpaused, revealing the transition pseudo-elements.
9. viewTransition.ready fulfills.
10. Pseudo-elements animate until finished.
11. Transition pseudo-elements removed.
12. viewTransition.finished fulfills.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-capture-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-start-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/group-animation-for-root-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/input-targets-root-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/mix-blend-mode-only-on-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-crash-set-exception-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-css-animation-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-raf-while-render-blocked-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-group-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-image-pair-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-new-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-old-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-on-root-element-with-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/only-child-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/paused-animation-at-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-computed-style-stays-in-sync-with-new-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/ready_resolves_after_dom_before_raf-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/style-inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/synchronous-callback-skipped-before-run-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-after-animation-started-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-from-invalid-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/unset-and-initial-view-transition-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-name-on-removed-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/web-animation-pseudo-incorrect-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document+ViewTransition.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::activeViewTransition const):
(WebCore::Document::setActiveViewTransition):
(WebCore::Document::hasViewTransitionPseudoElementTree const):
(WebCore::Document::setHasViewTransitionPseudoElementTree):
(WebCore::Document::startViewTransition):
(WebCore::Document::performPendingViewTransitions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.cpp: Added.
(WebCore::createPromiseAndWrapper):
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::ready):
(WebCore::ViewTransition::updateCallbackDone):
(WebCore::ViewTransition::finished):
(WebCore::ViewTransition::skipViewTransition):
(WebCore::ViewTransition::skipTransition):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::setupViewTransition):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::clearViewTransition):
* Source/WebCore/dom/ViewTransition.h: Added.
(WebCore::ViewTransition::phase):
(WebCore::ViewTransition::namedElements const):
(WebCore::ViewTransition::protectedDocument const):
* Source/WebCore/dom/ViewTransition.idl: Added.
* Source/WebCore/dom/ViewTransitionUpdateCallback.h: Added.
* Source/WebCore/dom/ViewTransitionUpdateCallback.idl: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator<<):
* Source/WebCore/page/Page.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):

Canonical link: https://commits.webkit.org/271426@main
